python_modules/libraries/dagster-aws/dagster_aws/pipes/message_readers.py

```diff
@@ -113,86 +133,171 @@ class CloudWatchEvent(TypedDict):
+            raise DagsterInvariantViolationError(
+                "log_group and log_stream must be set either in the constructor or in Pipes params."
+            )
+
+        self.thread = Thread(
+            target=self._start, kwargs={"params": params, "is_session_closed": is_session_closed}
         )
+        self.thread.start()
+
+    def _start(self, params: PipesParams, is_session_closed: Event) -> None:
+        log_group = cast(str, params.get("log_group") or self.log_group)
+        log_stream = cast(str, params.get("log_stream") or self.log_stream)
+        start_time = cast(int, self.start_time or params.get("start_time"))
 
-        for events_batch in self._get_all_cloudwatch_events(
-            log_group=log_group, log_stream=log_stream, start_time=start_time, end_time=end_time
+        for events in tail_cloudwatch_events(
+            self.client, log_group, log_stream, start_time=start_time
         ):
-            for event in events_batch:
-                for log_line in event["message"].splitlines():
-                    extract_message_or_forward_to_file(handler, log_line, output_file)
+            for event in events:
+                for line in event["message"].splitlines():
+                    forward_only_logs_to_file(line, self.target_stream)
 
-    def no_messages_debug_text(self) -> str:
-        return "Attempted to read messages by extracting them from CloudWatch logs directly."
+            if is_session_closed.is_set():
+                return
+
+    def stop(self) -> None:
+        pass
+
+    def is_running(self) -> bool:
+        return self.thread is not None and self.thread.is_alive()
 
-    def _get_all_cloudwatch_events(
+
+@experimental
+class PipesCloudWatchMessageReader(PipesThreadedMessageReader):
+    """Message reader that consumes AWS CloudWatch logs to read pipes messages."""
+
+    def __init__(
         self,
-        log_group: str,
-        log_stream: str,
-        start_time: Optional[int] = None,
-        end_time: Optional[int] = None,
-    ) -> Generator[List[CloudWatchEvent], None, None]:
-        """Returns batches of CloudWatch events until the stream is complete or end_time."""
-        params: Dict[str, Any] = {
-            "logGroupName": log_group,
-            "logStreamName": log_stream,
+        client=None,
+        log_group: Optional[str] = None,
+        log_stream: Optional[str] = None,
+        log_readers: Optional[Mapping[str, PipesLogReader]] = None,
+    ):
+        """Args:
+        client (boto3.client): boto3 CloudWatch client.
+        """
+        self.client: "LogsClient" = client or boto3.client("logs")
+        self.log_group = log_group
+        self.log_stream = log_stream
+
+        self.start_time = datetime.now()
+
+        super().__init__(log_readers=log_readers)
+
+    def on_launched(self, params: PipesLaunchedData) -> None:
+        if "log_group" in params["extras"]:
+            self.log_group = params["extras"]["log_group"]
```

Comment:
@smackesey seems like if we do something like this the `can_start` method doesn't even need to have `params` argument.

We can set the actual attributes in `on_opened` and `on_launched`. 

No need to introduce another source of parameters. 

What do you think?